hazelcast/src/main/java/com/hazelcast/spi/Offload.java

```diff
@@ -0,0 +1,142 @@
+ * is ready and can be returned by the operation.
+ *
+ * The Offload functionality automatically registers the source operation for operation heartbeats. The registration is done
+ * when the {@link Offload#init(NodeEngineImpl, Set)} is called. And it is unregistered, as soon as a response is send to the
+ * source operation.
+ *
+ * Todo:
+ * - Should an offload provide a guard against multiple responses?
+ * - Should a method be added to Offload that automatically reschedules the 'source' operation on an operation thread? It
+ * could also restore the original operationresponsehandler on the source operation.
+ */
+public abstract class Offload extends CallStatus {
+
+    protected InternalOperationService operationService;
+    protected NodeEngine nodeEngine;
+    protected ExecutionService executionService;
+    protected SerializationService serializationService;
+    private final Operation source;
+    private final OperationResponseHandler delegateResponseHandler;
+    private Set<Operation> asyncOperations;
+
+    /**
+     * Creates a new Offload instance.
+     *
+     * @param source the Operation that triggered this Offload to be created.
+     */
+    public Offload(final Operation source) {
+        super(OFFLOAD_ORDINAL);
+        this.source = source;
+        this.delegateResponseHandler = source.getOperationResponseHandler();
+    }
+
+    /**
+     * Returns the Operation that created this Offload. Returned operation should not be null.
+     *
+     * Currently this is use to automatically register the operation for the sake of heartbeats.
+     *
+     * @return the Operation.
+     */
+    public final Operation source() {
+        return source;
+    }
+
+    /**
+     * Initializes the Offload.
+     *
+     * As part of the initialization, the {@link OperationResponseHandler} of the source {@link Operation} is replaced by
+     * a decorated version that takes care of automatic deregistration of the operation on completion.
+     *
+     * This method is called before the {@link #start()} is called by the Operation infrastructure. An implementor of the
+     * {@link Offload} doesn't need to deal with this method.
+     *
+     * @param nodeEngine the {@link NodeEngineImpl}
+     */
+    public final void init(NodeEngineImpl nodeEngine, Set<Operation> asyncOperations) {
+        this.nodeEngine = nodeEngine;
+        this.operationService = nodeEngine.getOperationService();
+        this.serializationService = nodeEngine.getSerializationService();
+        this.asyncOperations = asyncOperations;
+        this.executionService = nodeEngine.getExecutionService();
+
+        asyncOperations.add(source);
+
+        if (shouldDecorateOperationResponseHandler()) {
+            // todo: it could happen that when an operation is offload multiple times, that you get a chain of offload
+            // response handlers.
+            source.setOperationResponseHandler(new OperationResponseHandlerImpl());
```

Comment:
Would be good to have a test for chaining. This can be tricky indeed.

Reply:
Yes. Currently we don't have true continuations (apart from BlockingOperations), so it isn't an issue so far.

Reply:
I have improved the documentation. It explicitly states that in case of an Operation that is offloaded, the response handler of the source operation should not be modified. 

I have not seen any violations of this. The response handlers of the forked operations are intercepted; but the response handler of the 'source' operation remains is not changed.